src/index.rs

```diff
@@ -231,8 +235,70 @@ impl IndexTable {
 		Ok(try_io!(Ok(&map[offset..offset + CHUNK_LEN])))
 	}
 
-	#[inline(never)]
 	fn find_entry(&self, key_prefix: u64, sub_index: usize, chunk: &[u8]) -> (Entry, usize) {
+		#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+		if is_x86_feature_detected!("sse2") {
+			return self.find_entry_sse2(key_prefix, sub_index, chunk)
+		}
+		self.find_entry_regular(key_prefix, sub_index, chunk)
+	}
+
+	#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+	fn find_entry_sse2(&self, key_prefix: u64, sub_index: usize, chunk: &[u8]) -> (Entry, usize) {
+		assert!(chunk.len() >= CHUNK_ENTRIES * 8); // Bound checking (not done by SIMD instructions)
+		debug_assert!(
+			Entry::address_bits(self.id.index_bits()) <= 32,
+			"To be sure we can use all high 32 bits as key prefix"
+		);
+		debug_assert_eq!(
+			CHUNK_ENTRIES % 4,
+			0,
+			"We assume here we got buffer with a number of elements that is a multiple of 4"
+		);
+
+		unsafe {
+			let target = _mm_set1_epi32(((key_prefix << self.id.index_bits()) >> 32) as i32);
+			let mut i = (sub_index >> 2) << 2; // We keep an alignment of 4
+			while i + 4 <= CHUNK_ENTRIES {
+				// We load the value 2 by 2 and move the high bits into the low part of the register
+				let first_two = _mm_shuffle_epi32::<0b10001101>(_mm_loadu_si128(
+					chunk[i * 8..].as_ptr() as *const __m128i,
+				));
+				let last_two = _mm_shuffle_epi32::<0b10001101>(_mm_loadu_si128(
+					chunk[(i + 2) * 8..].as_ptr() as *const __m128i,
+				));
+				// We set into current the input low parts in the interleaved order
+				let current = _mm_unpacklo_epi32(first_two, last_two);
```

Comment:
Maybe if we use `_mm_unpacklo_epi64` instead, the parts effectively would not be interleaved and it should be possible to use `leading_zeros()` with the mask?

Reply:
Thank you! That's much better! Done.